Launching a Site on Altis docs (#222)
humanmade/altis-documentation: other-docs/guides/launching-a-site-on-altis.md

# Launching a Site on Altis

The process for launching a site on Altis varies depending on the type of project. Nonetheless, there are common steps that need to be done no matter the context and this guide will walk you through what needs to be done in preparation for your site launch.

## Planning

> Create a launch issue in the project repository. This issue should include:

Reviewer: IMO, this should be a "we recommend", because we don't necessarily require people to open issues.

* Launch date
* Feature freeze date
* Code freeze date
* Content freeze date (if migrating an existing site)
* Team members needed for the deploy
* Domains hosted on Altis
* Domain(s) used for email

> * FROM address for each site in the network

Reviewer: Are these not the same thing?

Author: Not necessarily. There are cases that I am currently aware of where there are multiple domains, but the FROM address for all of them is a single email address. For example if the domains were:

``` 
 - happyfunland.com
 - happyfunlandnews.com
 - happyfunlandgames.com
 - cruisehappyfun.com
 - etc
```

and they wanted the From email from all of those to be `noreply@happyfunland.com`.

- domain.com
- domain2.com
- etc.

**What domain(s) will email need to be sent from?**
<!-- Enter as a list -->
- noreply@domain.com
- etc.

**For each site in the network, what will be the FROM email?
<!-- Enter as a list -->
- **Network Admin:** noreply@domain.com
- **domain.com:** some-email@domain.com
- etc.

## Environment
* **Database import link:**
* **Uploads import link:**
* **Who is providing the SSL certificate(s) (Altis or client)?:** 
* **Who controls the DNS for the client (Altis or client)?:**
* **Projected traffic:** 
* **Any other specifications or challenges?**
_Refer to the Altis [Limitations](https://docs.altis-dxp.com/cloud/limitations/) and [Page Caching](https://docs.altis-dxp.com/cloud/page-caching/) guides for specifics around what requests can be supported._

```

## Preparing Altis Cloud Environments
After you have submitted your support request in the Altis Dashboard, the Altis Support team will work on setting up your environments and be in touch with your team. When your new cloud environments are set up, you will be contacted and can begin deploying your code to those new environments.

> Once you have environments set up, you should restrict access to them by [requiring login](https://docs.altis-dxp.com/security/require-login/), forcing [PHP basic authentication](https://docs.altis-dxp.com/security/php-basic-auth/), or both. In your Altis configuration file (`composer.json`), add the following to require login both and PHP authentication. You may omit PHP authentication, but we recommend having at least one of these enabled for all non-production and pre-launch sites. You can override the require login setting to not require login on local environments. PHP authentication is disabled by default on local environments. Refer to the documentation pages for more information about Require Login and PHP Basic Authentication.

Reviewer: I would rephrase this to break it up a bit. Also, it's not "PHP authentication", it's "HTTP Basic authentication" (or just "Basic auth"); that might be something we need to correct in the other docs too?

```suggestion
Once you have environments set up, we recommend limiting access to your sites. By default, login will be required on your development and staging environments. Before your site launches, we also recommend requiring login for your production environment as well.

Logins can be enforced by [requiring user logins](https://docs.altis-dxp.com/security/require-login/), forcing [HTTP basic authentication](https://docs.altis-dxp.com/security/php-basic-auth/), or both.

In your Altis configuration file (`composer.json`), add the following to both require login and HTTP Basic authentication. You may omit Basic authentication, but we recommend having at least one of these enabled for all non-production and pre-launch sites. You can override the require login setting to not require login on local environments. Basic authentication is disabled by default on local environments. Refer to the documentation pages for more information about Require Login and Basic Authentication.
```

It seems to me like we should also just enable require-login on prod, so that when the site launches, it's easy to just remove those lines?

Author: > It seems to me like we should also just enable require-login on prod, so that when the site launches, it's easy to just remove those lines?

Could do. I can spin up an issue for it.

```json
{
	"extra": {
		"altis": {
			"modules": {
				"security": {
					"require-login": true
				}
			},
			"environments": {
				"local": {
					"modules": {
						"security": {
							"require-login": false
						}
					}
				}
			}
		}
	}
}
```

> 
> ## Post-launch

Reviewer: These steps are really just before the launch happens.

```suggestion

## Preparation for launch
```

# Launching a Site on Altis

The process for launching a site on Altis varies depending on the type of project. Nonetheless, there are common steps that need to be done no matter the context and this guide will walk you through what needs to be done in preparation for your site launch.

## Planning

Create a launch issue in the project repository. This issue should include:

* Client name
* Client business region

> * Client contact

Reviewer 2: This could be followed by an internal team for a given client in which case this isn't needed. Additionally I think now that this issue isn't something the product team would need to interact with directly this information should be implicit from both the project team and Altis side so we can drop it.

* Launch date
* Feature freeze date
* Code freeze date
* Content freeze date (if migrating an existing site)
* Team members needed for the deploy
* Domains hosted on Altis
* Domain(s) used for email
* FROM address for each site in the network

> * GitHub repository

Reviewer 2: If the issue is going on the project's own repo this should be implicit too.

Author: That might not be obvious if we're telling them to make the GitHub issue to gather information and then submit a Zendesk ticket from the Altis Dashboard. If we're not seeing this issue in GitHub, do you think we should still pull out the repo link?

* Deploy branch name per environment
* A launch checklist containing pre-launch, launch day and post-launch tasks (examples can be found below)
* Any additional notes or complexities for the project launch

An [example checklist](#Example-launch-checklist) for your GitHub issue can be found below. This issue is primarily for _internal project use_ and compiling all the necessary information. 

> Once the GitHub issue is created, this information can then be copied into [a support ticket in the Altis Dashboard](https://docs.altis-dxp.com/guides/getting-help-with-altis/) to alert the Altis Support Team.

Reviewer 2: Maybe offer some advice on the type of ticket and a suitable title, thinking it'll be a "Task" and maybe "Site Launch Preparation" as the title.

Author: Would they be able to use the Altis Dashboard to create a new site if one did not already exist or is that why we're assuming that the environments are already set up?

### Example launch checklist

```markdown
# Goal

<!-- A brief overview of the launch objective. Is it a migration? Is it a new site? -->

## Pre-launch checklist
<!-- To be filled out by the project team/lead. Not needed for Altis Support ticket. -->
- [ ] Are we deploying the correct branch?
- [ ] Has AWS SES been setup?
- [ ] Is a content freeze active?

> - [ ] Is the client ready for DNS switches?

Reviewer 2: Mentioning the client again but this might be the client doing the launch themselves, "is the site ready for the DNS switch over?"

Author: ```suggestion
- [ ] Is the site ready for the DNS switch?
```

> - [ ] Site is ready for HTTPS?

Reviewer 2: We're removing this right?

- [ ] Are any redirections needed?

## Domains hosted on Altis

> <!-- List of domains that will be on the stack -->

Reviewer 2: I think this just needs to be custom domains, so `<project>-staging.altis.cloud` and `<project>-dev.altis.cloud` are the default domains we'll provide for those environments.

Lets keep the focus on the production environment for this guide and issue template so remove the staging and development sections below.

### Production
- domain.com
- domain2.com
- etc.

### Staging
- domain-staging.altis.cloud
- domain2-staging.altis.cloud
- etc.

### Development
<!-- If applicable -->
- domain-dev.altis.cloud
- domain2-dev.altis.cloud
- etc.

**What domain(s) will email need to be sent from?**
<!-- Enter as a list -->
- noreply@domain.com
- etc.

**For each site in the network, what will be the FROM email?
<!-- Enter as a list -->
- **Network Admin:** noreply@domain.com
- **domain.com:** some-email@domain.com
- etc.

## Environment
* **GitHub repository:** https://github.com/{owner}/{repo}
* **GitHub repository deploy branches:**
  * **Production:** `main`
  * **Staging:** `staging`

>   * **Development:** `dev`

Reviewer 2: Shouldn't need this for the cloud team if we're assuming the production environment will exist already.

Author: My understanding is that the support ticket that this would be a part of would be to initially set up the environments. Is that not what you were thinking?

Author: Yep, need to assume we've already created the environments at this point and we're just prepping for launch. 👍

```

## Preparing Altis Cloud Environments
After you have submitted your support request in the Altis Dashboard, the Altis Support team will work on setting up your environments and be in touch with your team. When your new cloud environments are set up, you will be contacted and can begin deploying your code to those new environments.

Once you have environments set up, you should restrict access to them by [requiring login](https://docs.altis-dxp.com/security/require-login/), forcing [PHP basic authentication](https://docs.altis-dxp.com/security/php-basic-auth/), or both. In your Altis configuration file (`composer.json`), add the following to require login both and PHP authentication. You may omit PHP authentication, but we recommend having at least one of these enabled for all non-production and pre-launch sites. You can override the require login setting to not require login on local environments. PHP authentication is disabled by default on local environments. Refer to the documentation pages for more information about Require Login and PHP Basic Authentication.

```json

> "altis": {

Reviewer 2: I know it gets deeply nested but I'm trying to keep the example JSON configs complete so the full `{ "extra": { "altis" : { ... } } }`, at the very least this should be valid JSON so the code highlighting doesn't get confused.

	"modules": {
		"security": {
			"require-login": true,

> 			"php-basic-auth": {

Reviewer 2: Worth noting right now you can't use PHP Basic Auth on a production environment at all, at least not without messing around with the `hmauth_filter_dev_env` filter to add production as a dev environment. I'd just leave the example and advice as recommending requiring login for now.

Author: I forgot you had to filter to run it on production 👍
It's been a while since I looked at that code...

- Do the development or staging environments need to be synced?

## Site launch

Before the domain(s) are pointed to the Altis servers, a content migration should be performed to ensure the right content is loaded on the site when the switch is made active. When the most updated data has been imported into the database, a search-and-replace action should be performed so all the database entries that contain the URL of the site have been updated to the _correct_ URL. This can be done in the Altis Dashboard by either opening an SSH connection to the Sandbox Server or running a command on the WP CLI tab.

The following command would change all entries in the database containing `domain-production.altis.cloud` to `domain.com`. We recommend testing the change first using the `--dry-run` flag to make sure no unexpected tables are affected.

```bash
wp search-replace domain-production.altis.cloud domain.com --all-tables --network --url=domain-production.altis.cloud
```

This step will need to be repeated for any subdomains you have. Be sure to flush the cache when you are done, otherwise the old URLs will still be saved in the object cache.

```bash
wp cache flush
```

> After the database is updated and the site is ready to go, contact Altis Support to let them know that you are ready for the DNS switch to your new site.

Author: Anything else here? Do we need to specify contacting Altis Support via their launch ticket/Altis Dashboard vs Slack?

Reviewer 2: I think using the launch ticket makes sense to track the whole process and act as a log we can show. It makes sense to add that as a suggestion.

Author: @roborourke Is that not sufficiently covered [here](https://github.com/humanmade/altis-documentation/pull/222/files#diff-6cf7fa890a1e26dae62794924a67bdaf3f4bd7fee7a8f72479a6f4a4539e5ac6R23)?

Reviewer 2: Yeah true, staring at it too long!

# Launching a Site on Altis

The process for launching a site on Altis varies depending on the type of project. Nonetheless, there are common steps that need to be done no matter the context and this guide will walk you through what needs to be done in preparation for your site launch.

## Planning

> Create a launch issue in the project repository. This issue should include:

Reviewer 2: We should note advice on setting require login to true for the production environment until after the domains and everything have been setup and migrated to, then as a last step, require login should be removed, committed and deployed

## Example launch checklist

```markdown
# Goal

<!-- A brief overview of the launch objective. Is it a migration? Is it a new site? -->

## Pre-launch checklist
<!-- To be filled out by the project team/lead. Not needed for Altis Support ticket. -->
- [ ] Are we deploying the correct branch?
- [ ] Has AWS SES been setup?
- [ ] Is a content freeze active?
- [ ] Is the client ready for DNS switches?
- [ ] Site is ready for HTTPS?
- [ ] Are any redirections needed?

## Domains hosted on Altis
<!-- List of domains that will be on the stack -->

### Production
- domain.com
- domain2.com
- etc.

### Staging
- domain-staging.altis.cloud
- domain2-staging.altis.cloud
- etc.

### Development
- domain-dev.altis.cloud
- domain2-dev.altis.cloud
- etc.

**What domain(s) will email need to be sent from?**
<!-- Enter as a list -->
- [ ] noreply@domain.com
- [ ] etc.

**For each site in the network, what will be the FROM email?
<!-- Enter as a list -->
- [ ] **Network Admin:** noreply@domain.com
- [ ] **domain.com:** some-email@domain.com
- [ ] etc.

## Environment

> * **GitHub repository:** https://github.com/humanmade/{projectname}

Reviewer 2: ```suggestion
* **GitHub repository:** https://github.com/{owner}/{repo}
```

* **GitHub repository deploy branches:**
  * **Production:** `main`
  * **Staging:** `staging`
  * **Development:** `dev`

>   * **Altis integration branch:** `altis` <!-- If applicable -->

Reviewer 2: Seems to be more a migration specific thing so we could dump this

* [ ] **Database exists in Altis Dashboard?**
  * [ ] If not, provide a link to the database to import:
* [ ] **Uploads exist in Altis Dashboard?**

>   * [ ] If not, provide a link to the uploads to import:  

Reviewer 2: Update to just request this information eg. the where, if it exists - need to check whether we actually do the initial import / transfer

* **Who is providing the SSL certificate(s) (Altis or client)?:** 
* **Who controls the DNS for the client (Altis or client)?:**

> * **Special requirements or considerations for ElasticSearch:**

Reviewer 2: Let's drop this! We don't want to invite any requests for setting snowflakey environments

* **Projected traffic:** 
* **Specific cookies that need to be allow-listed outside these patterns `comment_`, `hm_`, `wordpress_`, `wp-`, `wp_`, `altis_`:**

> * **Any other specifications or challenges?**

Reviewer 2: Make a reference the cloud env limitations doc

## Altis Cloud
- [ ] Nameserver records have been provided to the client (if we're managing their DNS).
- [ ] AWS SES domain verification DNS records have been provided to the client or entered into Route 53.
- [ ] AWS SES Sending domains have been verified in SES.
- [ ] AWS SES Sending domains have been configured in the application code.
- [ ] ACM Certificate DNS records have been provided to the client or entered into Route 53
- [ ] ACM Certificate has been issued.
- [ ] ACM Certificate has been attached to the environment.
- [ ] Existing database, if any, has been imported into the environment.
- [ ] Existing uploads, if any, have been imported into the environment.
- [ ] Database has been search/replaced with the environment-specific domain as the replacement.

<!-- For each environment that needs to be created or migrated, include a checklist like the following: -->

> ### domain-dev

Reviewer 2: These sections should just be internal docs replaced by a note to say that once the zendesk ticket has been created a cloud engineer will respond to that ticket with the following information:

specific info depending on some of the above, eg. are we managing their DNS

Author: Added a new section between submitting a support ticket and launch that sort of covers this a bit ("the Altis Support team will work on setting up your environments and be in touch with your team"). Let me know if you think anything additional is needed here.

- [ ] Communicate an edit lock with the client.
- [ ] Spin up the Terraform stack.
- [ ] Verify the stack is functioning correctly
- [ ] Point Terraform to the original CloudFront distribution
- [ ] Remove the existing distribution from Terraform's state
- [ ] Import the original distribution
- [ ] Update CloudFront domains with the values from the original distribution
- [ ] Plan the changes:
- [ ] Verify the changes looks
- [ ] Inform client edit lock has begun.
- [ ] Prepare to flip the switch
- [ ] Flip the switch:
- [ ] Monitor the RequestCount metric on the original Load Balancer
- [ ] Disable Cavalcade on the CloudFormation stack
- [ ] Verify the new RDS Cluster is in sync with the Original RDS Instance
- [ ] Disable Replication
- [ ] Enable writes on the new infrastructure
- [ ] Delete the dangling CloudFront distribution.
- [ ] Update the CloudFormation stack template to retain CloudFront distribution.
- [ ] Delete Cloudformation stack once after the client QA.

```

## Post-launch

After the site launch is completed, there may be some items left to clean up or check.

> - Do you need to set the default branch back to `main` in your repository?

Reviewer 2: Might be related to the migration branch note above? Not sure this is necessary

Author: That's probably accurate. New sites wouldn't likely need to switch deploy branches.
